// tdesign-vue-next/es/_chunks/dep-40c5ec80.mjs
/**
 * tdesign v1.15.2
 * (c) 2025 tdesign
 * @license MIT
 */

import { _ as _typeof$2 } from './dep-15464fee.mjs';
import { g as getDefaultExportFromCjs } from './dep-779bddf7.mjs';

function asyncGeneratorStep(n, t, e, r, o, a, c) {
  try {
    var i = n[a](c),
      u = i.value;
  } catch (n) {
    return void e(n);
  }
  i.done ? t(u) : Promise.resolve(u).then(r, o);
}
function _asyncToGenerator(n) {
  return function () {
    var t = this,
      e = arguments;
    return new Promise(function (r, o) {
      var a = n.apply(t, e);
      function _next(n) {
        asyncGeneratorStep(a, r, o, _next, _throw, "next", n);
      }
      function _throw(n) {
        asyncGeneratorStep(a, r, o, _next, _throw, "throw", n);
      }
      _next(void 0);
    });
  };
}

var regeneratorRuntime$3 = {exports: {}};

var OverloadYield$2 = {exports: {}};

var OverloadYield = OverloadYield$2.exports;
(function (module) {
  function _OverloadYield(e, d) {
    this.v = e, this.k = d;
  }
  module.exports = _OverloadYield, module.exports.__esModule = true, module.exports["default"] = module.exports;
})(OverloadYield$2);
var OverloadYieldExports = OverloadYield$2.exports;
var OverloadYield$1 = /*@__PURE__*/getDefaultExportFromCjs(OverloadYieldExports);

var regenerator$3 = {exports: {}};

var regeneratorDefine$2 = {exports: {}};

var regeneratorDefine = regeneratorDefine$2.exports;
(function (module) {
  function _regeneratorDefine(e, r, n, t) {
    var i = Object.defineProperty;
    try {
      i({}, "", {});
    } catch (e) {
      i = 0;
    }
    module.exports = _regeneratorDefine = function regeneratorDefine(e, r, n, t) {
      if (r) i ? i(e, r, {
        value: n,
        enumerable: !t,
        configurable: !t,
        writable: !t
      }) : e[r] = n;else {
        var o = function o(r, n) {
          _regeneratorDefine(e, r, function (e) {
            return this._invoke(r, n, e);
          });
        };
        o("next", 0), o("throw", 1), o("return", 2);
      }
    }, module.exports.__esModule = true, module.exports["default"] = module.exports, _regeneratorDefine(e, r, n, t);
  }
  module.exports = _regeneratorDefine, module.exports.__esModule = true, module.exports["default"] = module.exports;
})(regeneratorDefine$2);
var regeneratorDefineExports = regeneratorDefine$2.exports;
var regeneratorDefine$1 = /*@__PURE__*/getDefaultExportFromCjs(regeneratorDefineExports);

var regenerator$1 = regenerator$3.exports;
(function (module) {
  var regeneratorDefine = regeneratorDefineExports;
  function _regenerator() {
    /*! regenerator-runtime -- Copyright (c) 2014-present, Facebook, Inc. -- license (MIT): https://github.com/babel/babel/blob/main/packages/babel-helpers/LICENSE */
    var e,
      t,
      r = "function" == typeof Symbol ? Symbol : {},
      n = r.iterator || "@@iterator",
      o = r.toStringTag || "@@toStringTag";
    function i(r, n, o, i) {
      var c = n && n.prototype instanceof Generator ? n : Generator,
        u = Object.create(c.prototype);
      return regeneratorDefine(u, "_invoke", function (r, n, o) {
        var i,
          c,
          u,
          f = 0,
          p = o || [],
          y = !1,
          G = {
            p: 0,
            n: 0,
            v: e,
            a: d,
            f: d.bind(e, 4),
            d: function d(t, r) {
              return i = t, c = 0, u = e, G.n = r, a;
            }
          };
        function d(r, n) {
          for (c = r, u = n, t = 0; !y && f && !o && t < p.length; t++) {
            var o,
              i = p[t],
              d = G.p,
              l = i[2];
            r > 3 ? (o = l === n) && (u = i[(c = i[4]) ? 5 : (c = 3, 3)], i[4] = i[5] = e) : i[0] <= d && ((o = r < 2 && d < i[1]) ? (c = 0, G.v = n, G.n = i[1]) : d < l && (o = r < 3 || i[0] > n || n > l) && (i[4] = r, i[5] = n, G.n = l, c = 0));
          }
          if (o || r > 1) return a;
          throw y = !0, n;
        }
        return function (o, p, l) {
          if (f > 1) throw TypeError("Generator is already running");
          for (y && 1 === p && d(p, l), c = p, u = l; (t = c < 2 ? e : u) || !y;) {
            i || (c ? c < 3 ? (c > 1 && (G.n = -1), d(c, u)) : G.n = u : G.v = u);
            try {
              if (f = 2, i) {
                if (c || (o = "next"), t = i[o]) {
                  if (!(t = t.call(i, u))) throw TypeError("iterator result is not an object");
                  if (!t.done) return t;
                  u = t.value, c < 2 && (c = 0);
                } else 1 === c && (t = i["return"]) && t.call(i), c < 2 && (u = TypeError("The iterator does not provide a '" + o + "' method"), c = 1);
                i = e;
              } else if ((t = (y = G.n < 0) ? u : r.call(n, G)) !== a) break;
            } catch (t) {
              i = e, c = 1, u = t;
            } finally {
              f = 1;
            }
          }
          return {
            value: t,
            done: y
          };
        };
      }(r, o, i), !0), u;
    }
    var a = {};
    function Generator() {}
    function GeneratorFunction() {}
    function GeneratorFunctionPrototype() {}
    t = Object.getPrototypeOf;
    var c = [][n] ? t(t([][n]())) : (regeneratorDefine(t = {}, n, function () {
        return this;
      }), t),
      u = GeneratorFunctionPrototype.prototype = Generator.prototype = Object.create(c);
    function f(e) {
      return Object.setPrototypeOf ? Object.setPrototypeOf(e, GeneratorFunctionPrototype) : (e.__proto__ = GeneratorFunctionPrototype, regeneratorDefine(e, o, "GeneratorFunction")), e.prototype = Object.create(u), e;
    }
    return GeneratorFunction.prototype = GeneratorFunctionPrototype, regeneratorDefine(u, "constructor", GeneratorFunctionPrototype), regeneratorDefine(GeneratorFunctionPrototype, "constructor", GeneratorFunction), GeneratorFunction.displayName = "GeneratorFunction", regeneratorDefine(GeneratorFunctionPrototype, o, "GeneratorFunction"), regeneratorDefine(u), regeneratorDefine(u, o, "Generator"), regeneratorDefine(u, n, function () {
      return this;
    }), regeneratorDefine(u, "toString", function () {
      return "[object Generator]";
    }), (module.exports = _regenerator = function _regenerator() {
      return {
        w: i,
        m: f
      };
    }, module.exports.__esModule = true, module.exports["default"] = module.exports)();
  }
  module.exports = _regenerator, module.exports.__esModule = true, module.exports["default"] = module.exports;
})(regenerator$3);
var regeneratorExports = regenerator$3.exports;
var regenerator$2 = /*@__PURE__*/getDefaultExportFromCjs(regeneratorExports);

var regeneratorAsync$2 = {exports: {}};

var regeneratorAsyncGen$2 = {exports: {}};

var regeneratorAsyncIterator$2 = {exports: {}};

var regeneratorAsyncIterator = regeneratorAsyncIterator$2.exports;
(function (module) {
  var OverloadYield = OverloadYieldExports;
  var regeneratorDefine = regeneratorDefineExports;
  function AsyncIterator(t, e) {
    function n(r, o, i, f) {
      try {
        var c = t[r](o),
          u = c.value;
        return u instanceof OverloadYield ? e.resolve(u.v).then(function (t) {
          n("next", t, i, f);
        }, function (t) {
          n("throw", t, i, f);
        }) : e.resolve(u).then(function (t) {
          c.value = t, i(c);
        }, function (t) {
          return n("throw", t, i, f);
        });
      } catch (t) {
        f(t);
      }
    }
    var r;
    this.next || (regeneratorDefine(AsyncIterator.prototype), regeneratorDefine(AsyncIterator.prototype, "function" == typeof Symbol && Symbol.asyncIterator || "@asyncIterator", function () {
      return this;
    })), regeneratorDefine(this, "_invoke", function (t, o, i) {
      function f() {
        return new e(function (e, r) {
          n(t, i, e, r);
        });
      }
      return r = r ? r.then(f, f) : f();
    }, !0);
  }
  module.exports = AsyncIterator, module.exports.__esModule = true, module.exports["default"] = module.exports;
})(regeneratorAsyncIterator$2);
var regeneratorAsyncIteratorExports = regeneratorAsyncIterator$2.exports;
var regeneratorAsyncIterator$1 = /*@__PURE__*/getDefaultExportFromCjs(regeneratorAsyncIteratorExports);

var regeneratorAsyncGen = regeneratorAsyncGen$2.exports;
(function (module) {
  var regenerator = regeneratorExports;
  var regeneratorAsyncIterator = regeneratorAsyncIteratorExports;
  function _regeneratorAsyncGen(r, e, t, o, n) {
    return new regeneratorAsyncIterator(regenerator().w(r, e, t, o), n || Promise);
  }
  module.exports = _regeneratorAsyncGen, module.exports.__esModule = true, module.exports["default"] = module.exports;
})(regeneratorAsyncGen$2);
var regeneratorAsyncGenExports = regeneratorAsyncGen$2.exports;
var regeneratorAsyncGen$1 = /*@__PURE__*/getDefaultExportFromCjs(regeneratorAsyncGenExports);

var regeneratorAsync = regeneratorAsync$2.exports;
(function (module) {
  var regeneratorAsyncGen = regeneratorAsyncGenExports;
  function _regeneratorAsync(n, e, r, t, o) {
    var a = regeneratorAsyncGen(n, e, r, t, o);
    return a.next().then(function (n) {
      return n.done ? n.value : a.next();
    });
  }
  module.exports = _regeneratorAsync, module.exports.__esModule = true, module.exports["default"] = module.exports;
})(regeneratorAsync$2);
var regeneratorAsyncExports = regeneratorAsync$2.exports;
var regeneratorAsync$1 = /*@__PURE__*/getDefaultExportFromCjs(regeneratorAsyncExports);

var regeneratorKeys$2 = {exports: {}};

var regeneratorKeys = regeneratorKeys$2.exports;
(function (module) {
  function _regeneratorKeys(e) {
    var n = Object(e),
      r = [];
    for (var t in n) r.unshift(t);
    return function e() {
      for (; r.length;) if ((t = r.pop()) in n) return e.value = t, e.done = !1, e;
      return e.done = !0, e;
    };
  }
  module.exports = _regeneratorKeys, module.exports.__esModule = true, module.exports["default"] = module.exports;
})(regeneratorKeys$2);
var regeneratorKeysExports = regeneratorKeys$2.exports;
var regeneratorKeys$1 = /*@__PURE__*/getDefaultExportFromCjs(regeneratorKeysExports);

var regeneratorValues$2 = {exports: {}};

var _typeof$1 = {exports: {}};

var _typeof_1 = _typeof$1.exports;
(function (module) {
  function _typeof(o) {
    "@babel/helpers - typeof";

    return module.exports = _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) {
      return typeof o;
    } : function (o) {
      return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o;
    }, module.exports.__esModule = true, module.exports["default"] = module.exports, _typeof(o);
  }
  module.exports = _typeof, module.exports.__esModule = true, module.exports["default"] = module.exports;
})(_typeof$1);
var _typeofExports = _typeof$1.exports;
var _typeof = /*@__PURE__*/getDefaultExportFromCjs(_typeofExports);

var regeneratorValues = regeneratorValues$2.exports;
(function (module) {
  var _typeof = _typeofExports["default"];
  function _regeneratorValues(e) {
    if (null != e) {
      var t = e["function" == typeof Symbol && Symbol.iterator || "@@iterator"],
        r = 0;
      if (t) return t.call(e);
      if ("function" == typeof e.next) return e;
      if (!isNaN(e.length)) return {
        next: function next() {
          return e && r >= e.length && (e = void 0), {
            value: e && e[r++],
            done: !e
          };
        }
      };
    }
    throw new TypeError(_typeof(e) + " is not iterable");
  }
  module.exports = _regeneratorValues, module.exports.__esModule = true, module.exports["default"] = module.exports;
})(regeneratorValues$2);
var regeneratorValuesExports = regeneratorValues$2.exports;
var regeneratorValues$1 = /*@__PURE__*/getDefaultExportFromCjs(regeneratorValuesExports);

var regeneratorRuntime$1 = regeneratorRuntime$3.exports;
(function (module) {
  var OverloadYield = OverloadYieldExports;
  var regenerator = regeneratorExports;
  var regeneratorAsync = regeneratorAsyncExports;
  var regeneratorAsyncGen = regeneratorAsyncGenExports;
  var regeneratorAsyncIterator = regeneratorAsyncIteratorExports;
  var regeneratorKeys = regeneratorKeysExports;
  var regeneratorValues = regeneratorValuesExports;
  function _regeneratorRuntime() {
    "use strict";

    var r = regenerator(),
      e = r.m(_regeneratorRuntime),
      t = (Object.getPrototypeOf ? Object.getPrototypeOf(e) : e.__proto__).constructor;
    function n(r) {
      var e = "function" == typeof r && r.constructor;
      return !!e && (e === t || "GeneratorFunction" === (e.displayName || e.name));
    }
    var o = {
      "throw": 1,
      "return": 2,
      "break": 3,
      "continue": 3
    };
    function a(r) {
      var e, t;
      return function (n) {
        e || (e = {
          stop: function stop() {
            return t(n.a, 2);
          },
          "catch": function _catch() {
            return n.v;
          },
          abrupt: function abrupt(r, e) {
            return t(n.a, o[r], e);
          },
          delegateYield: function delegateYield(r, o, a) {
            return e.resultName = o, t(n.d, regeneratorValues(r), a);
          },
          finish: function finish(r) {
            return t(n.f, r);
          }
        }, t = function t(r, _t, o) {
          n.p = e.prev, n.n = e.next;
          try {
            return r(_t, o);
          } finally {
            e.next = n.n;
          }
        }), e.resultName && (e[e.resultName] = n.v, e.resultName = void 0), e.sent = n.v, e.next = n.n;
        try {
          return r.call(this, e);
        } finally {
          n.p = e.prev, n.n = e.next;
        }
      };
    }
    return (module.exports = _regeneratorRuntime = function _regeneratorRuntime() {
      return {
        wrap: function wrap(e, t, n, o) {
          return r.w(a(e), t, n, o && o.reverse());
        },
        isGeneratorFunction: n,
        mark: r.m,
        awrap: function awrap(r, e) {
          return new OverloadYield(r, e);
        },
        AsyncIterator: regeneratorAsyncIterator,
        async: function async(r, e, t, o, u) {
          return (n(e) ? regeneratorAsyncGen : regeneratorAsync)(a(r), e, t, o, u);
        },
        keys: regeneratorKeys,
        values: regeneratorValues
      };
    }, module.exports.__esModule = true, module.exports["default"] = module.exports)();
  }
  module.exports = _regeneratorRuntime, module.exports.__esModule = true, module.exports["default"] = module.exports;
})(regeneratorRuntime$3);
var regeneratorRuntimeExports = regeneratorRuntime$3.exports;
var regeneratorRuntime$2 = /*@__PURE__*/getDefaultExportFromCjs(regeneratorRuntimeExports);

// TODO(Babel 8): Remove this file.

var runtime = regeneratorRuntimeExports();
var regenerator = runtime;

// Copied from https://github.com/facebook/regenerator/blob/main/packages/runtime/runtime.js#L736=
try {
  regeneratorRuntime = runtime;
} catch (accidentalStrictMode) {
  if ((typeof globalThis === "undefined" ? "undefined" : _typeof$2(globalThis)) === "object") {
    globalThis.regeneratorRuntime = runtime;
  } else {
    Function("r", "regeneratorRuntime = r")(runtime);
  }
}
var _regeneratorRuntime = /*@__PURE__*/getDefaultExportFromCjs(regenerator);

export { _asyncToGenerator as _, _regeneratorRuntime as a };
