// tdesign-vue-next/esm/_chunks/dep-262abab0.js
/**
 * tdesign v1.15.2
 * (c) 2025 tdesign
 * @license MIT
 */

var TD_COLOR_USED_COLORS_MAX_SIZE = 100;
var COLOR_MODES = {
  monochrome: "singleColor",
  "linear-gradient": "gradientColor"
};
var DEFAULT_COLOR = "#001F97";
var DEFAULT_LINEAR_GRADIENT = "linear-gradient(90deg, rgba(241,29,0,1) 0%, rgba(73,106,220,1) 100%);";
var DEFAULT_SYSTEM_SWATCH_COLORS = ["#ECF2FE", "#D4E3FC", "#BBD3FB", "#96BBF8", "#699EF5", "#4787F0", "#266FE8", "#0052D9", "#0034B5", "#001F97", "#FDECEE", "#F9D7D9", "#F8B9BE", "#F78D94", "#F36D78", "#E34D59", "#C9353F", "#B11F26", "#951114", "#680506", "#FEF3E6", "#F9E0C7", "#F7C797", "#F2995F", "#ED7B2F", "#D35A21", "#BA431B", "#9E3610", "#842B0B", "#5A1907", "#E8F8F2", "#BCEBDC", "#85DBBE", "#48C79C", "#00A870", "#078D5C", "#067945", "#056334", "#044F2A", "#033017"];
var FORMATS = ["HEX", "RGB", "HSL", "HSV", "CMYK", "CSS"];
var ALPHA_FORMAT_MAP = {
  HEX: "HEX8",
  RGB: "RGBA",
  HSL: "HSLA",
  HSV: "HSVA"
};
var COLOR_FORMAT_INPUTS = {
  RGB: [{
    key: "r",
    min: 0,
    max: 255,
    type: "inputNumber"
  }, {
    key: "g",
    min: 0,
    max: 255,
    type: "inputNumber"
  }, {
    key: "b",
    min: 0,
    max: 255,
    type: "inputNumber"
  }],
  HSV: [{
    key: "h",
    min: 0,
    max: 360,
    type: "inputNumber"
  }, {
    key: "s",
    min: 0,
    max: 100,
    type: "inputNumber"
  }, {
    key: "v",
    min: 0,
    max: 100,
    type: "inputNumber"
  }],
  HSL: [{
    key: "h",
    min: 0,
    max: 360,
    type: "inputNumber"
  }, {
    key: "s",
    min: 0,
    max: 100,
    type: "inputNumber"
  }, {
    key: "l",
    min: 0,
    max: 100,
    type: "inputNumber"
  }],
  CMYK: [{
    key: "c",
    min: 0,
    max: 255,
    type: "inputNumber"
  }, {
    key: "m",
    min: 0,
    max: 255,
    type: "inputNumber"
  }, {
    key: "y",
    min: 0,
    max: 255,
    type: "inputNumber"
  }, {
    key: "k",
    min: 0,
    max: 255,
    type: "inputNumber"
  }],
  CSS: [{
    key: "css",
    type: "input",
    flex: 3
  }],
  HEX: [{
    key: "hex",
    type: "input",
    flex: 3
  }]
};
var SATURATION_PANEL_DEFAULT_WIDTH = 230;
var SATURATION_PANEL_DEFAULT_HEIGHT = 168;
var SLIDER_DEFAULT_WIDTH = 186;
var GRADIENT_SLIDER_DEFAULT_WIDTH = 176;

export { ALPHA_FORMAT_MAP as A, COLOR_FORMAT_INPUTS as C, DEFAULT_LINEAR_GRADIENT as D, FORMATS as F, GRADIENT_SLIDER_DEFAULT_WIDTH as G, SATURATION_PANEL_DEFAULT_WIDTH as S, TD_COLOR_USED_COLORS_MAX_SIZE as T, DEFAULT_COLOR as a, DEFAULT_SYSTEM_SWATCH_COLORS as b, COLOR_MODES as c, SATURATION_PANEL_DEFAULT_HEIGHT as d, SLIDER_DEFAULT_WIDTH as e };
