// tdesign-vue-next/es/index.mjs
/**
 * tdesign v1.15.2
 * (c) 2025 tdesign
 * @license MIT
 */

import { c as components } from './_chunks/dep-4946596e.mjs';
import './common.mjs';
export { Button } from './button/index.mjs';
export { Icon } from './icon/index.mjs';
export { Link } from './link/index.mjs';
export { Col, Row } from './grid/index.mjs';
export { Aside, Content, Footer, Header, Layout } from './layout/index.mjs';
export { Divider } from './divider/index.mjs';
export { Space } from './space/index.mjs';
export { Descriptions, DescriptionsItem } from './descriptions/index.mjs';
export { Affix } from './affix/index.mjs';
export { Anchor, AnchorItem, AnchorTarget } from './anchor/index.mjs';
export { Breadcrumb, BreadcrumbItem } from './breadcrumb/index.mjs';
export { Dropdown, DropdownItem, DropdownMenu } from './dropdown/index.mjs';
export { HeadMenu, Menu, MenuGroup, MenuItem, Submenu } from './menu/index.mjs';
export { Pagination, PaginationMini } from './pagination/index.mjs';
export { StepItem, Steps } from './steps/index.mjs';
export { TabPanel, Tabs } from './tabs/index.mjs';
export { StickyItem, StickyTool } from './sticky-tool/index.mjs';
export { BackTop } from './back-top/index.mjs';
export { AutoComplete, HighlightOption } from './auto-complete/index.mjs';
export { Cascader, CascaderPanel } from './cascader/index.mjs';
export { Checkbox, CheckboxGroup } from './checkbox/index.mjs';
export { DatePicker, DatePickerPanel, DateRangePicker, DateRangePickerPanel } from './date-picker/index.mjs';
export { Form, FormItem } from './form/index.mjs';
export { InputAdornment } from './input-adornment/index.mjs';
export { Input, InputGroup } from './input/index.mjs';
export { InputNumber } from './input-number/index.mjs';
export { Radio, RadioButton, RadioGroup } from './radio/index.mjs';
export { Option, OptionGroup, Select } from './select/index.mjs';
export { Slider, SliderButton } from './slider/index.mjs';
export { Switch } from './switch/index.mjs';
export { TagInput } from './tag-input/index.mjs';
export { SelectInput } from './select-input/index.mjs';
export { Textarea } from './textarea/index.mjs';
export { Transfer } from './transfer/index.mjs';
export { TimePicker, TimePickerPanel, TimeRangePicker } from './time-picker/index.mjs';
export { TreeSelect } from './tree-select/index.mjs';
export { ColorPicker, ColorPickerPanel } from './color-picker/index.mjs';
export { RangeInput, RangeInputPopup } from './range-input/index.mjs';
export { Avatar, AvatarGroup } from './avatar/index.mjs';
export { Badge } from './badge/index.mjs';
export { Calendar } from './calendar/index.mjs';
export { Card } from './card/index.mjs';
export { Comment } from './comment/index.mjs';
export { Empty } from './empty/index.mjs';
export { Image } from './image/index.mjs';
export { ImageViewer } from './image-viewer/index.mjs';
export { List, ListItem, ListItemMeta } from './list/index.mjs';
export { Progress } from './progress/index.mjs';
export { Skeleton } from './skeleton/index.mjs';
export { Statistic } from './statistic/index.mjs';
export { Swiper, SwiperItem } from './swiper/index.mjs';
export { BaseTable, EnhancedTable, PrimaryTable, Table } from './table/index.mjs';
export { CheckTag, CheckTagGroup, Tag } from './tag/index.mjs';
export { Timeline, TimelineItem } from './timeline/index.mjs';
export { Tooltip } from './tooltip/index.mjs';
export { Tree } from './tree/index.mjs';
export { Collapse, CollapsePanel } from './collapse/index.mjs';
export { Watermark } from './watermark/index.mjs';
export { Rate } from './rate/index.mjs';
export { Paragraph, Text, Title, Typography } from './typography/index.mjs';
export { QRCode } from './qrcode/index.mjs';
export { Alert } from './alert/index.mjs';
export { DialogPlugin } from './dialog/plugin.mjs';
export { Dialog, DialogCard } from './dialog/index.mjs';
export { DrawerPlugin } from './drawer/plugin.mjs';
export { Drawer } from './drawer/index.mjs';
export { Guide } from './guide/index.mjs';
export { LoadingPlugin } from './loading/plugin.mjs';
export { vLoading as LoadingDirective } from './loading/directive.mjs';
export { Loading } from './loading/index.mjs';
export { MessagePlugin } from './message/plugin.mjs';
export { Message } from './message/index.mjs';
export { default as NotifyPlugin } from './notification/plugin.mjs';
export { Notification } from './notification/index.mjs';
export { Popconfirm } from './popconfirm/index.mjs';
export { Popup } from './popup/index.mjs';
export { Upload } from './upload/index.mjs';
export { ConfigProvider } from './config-provider/index.mjs';
import './button/button.mjs';
import './_chunks/dep-d0add92f.mjs';
import './_chunks/dep-32b59907.mjs';
import 'vue';
import './_chunks/dep-0bd8597f.mjs';
import './_chunks/dep-612a2c2b.mjs';
import './_chunks/dep-15464fee.mjs';
import './button/props.mjs';
import './_chunks/dep-465c43e8.mjs';
import 'lodash-es';
import './_chunks/dep-d58b61b6.mjs';
import './_chunks/dep-779bddf7.mjs';
import './config-provider/hooks/useConfig.mjs';
import './config-provider/utils/context.mjs';
import './_chunks/dep-afae046d.mjs';
import './_chunks/dep-8a6c1499.mjs';
import './_chunks/dep-5c28ada1.mjs';
import './_chunks/dep-614f307d.mjs';
import './_chunks/dep-d913bc66.mjs';
import './_chunks/dep-1690abc9.mjs';
import './_chunks/dep-62c11543.mjs';
import './_chunks/dep-27c2b283.mjs';
import './_chunks/dep-67238d91.mjs';
import './loading/loading.mjs';
import './loading/icon/gradient.mjs';
import './_chunks/dep-6b6765a0.mjs';
import './_chunks/dep-3ea2b330.mjs';
import './loading/props.mjs';
import './style/css.mjs';
import 'tdesign-icons-vue-next';
import './link/link.mjs';
import './link/props.mjs';
import './grid/row.mjs';
import './grid/row-props.mjs';
import './grid/hooks/index.mjs';
import './grid/utils/index.mjs';
import './grid/col.mjs';
import './grid/col-props.mjs';
import './layout/layout.mjs';
import './layout/header.mjs';
import './layout/header-props.mjs';
import './layout/footer.mjs';
import './layout/footer-props.mjs';
import './layout/aside.mjs';
import './layout/aside-props.mjs';
import './layout/content.mjs';
import './layout/content-props.mjs';
import './divider/divider.mjs';
import './divider/props.mjs';
import './space/space.mjs';
import './space/props.mjs';
import './descriptions/descriptions.mjs';
import './descriptions/props.mjs';
import './descriptions/consts/index.mjs';
import './descriptions/descriptions-row.mjs';
import './_chunks/dep-8269a1cd.mjs';
import './descriptions/descriptions-item.mjs';
import './descriptions/descriptions-item-props.mjs';
import './affix/affix.mjs';
import './_chunks/dep-40c5ec80.mjs';
import './affix/props.mjs';
import './anchor/anchor.mjs';
import './anchor/utils/index.mjs';
import './anchor/props.mjs';
import './anchor/consts/index.mjs';
import './anchor/anchor-item.mjs';
import './anchor/anchor-item-props.mjs';
import './anchor/anchor-target.mjs';
import './anchor/anchor-target-props.mjs';
import './message/message-list.mjs';
import './message/consts/index.mjs';
import './message/message.mjs';
import './message/props.mjs';
import './message/utils/index.mjs';
import './message/utils/animation.mjs';
import './popup/popup.mjs';
import '@popperjs/core';
import './popup/container.mjs';
import './popup/props.mjs';
import './breadcrumb/breadcrumb.mjs';
import './breadcrumb/props.mjs';
import './breadcrumb/breadcrumb-item.mjs';
import './breadcrumb/breadcrumb-item-props.mjs';
import './tooltip/tooltip.mjs';
import './tooltip/props.mjs';
import './tooltip/utils/index.mjs';
import './breadcrumb/hooks/index.mjs';
import './breadcrumb/hooks/useEllipsis.mjs';
import './_chunks/dep-6686ef20.mjs';
import './breadcrumb/hooks/useBreadcrumbOptions.mjs';
import './dropdown/dropdown-item.mjs';
import './dropdown/dropdown-item-props.mjs';
import './dropdown/dropdown.mjs';
import './dropdown/dropdown-menu.mjs';
import './dropdown/props.mjs';
import './dropdown/hooks/useDropdownOptions.mjs';
import './menu/menu.mjs';
import './menu/props.mjs';
import './menu/utils/index.mjs';
import './menu/utils/v-menu.mjs';
import './menu/head-menu.mjs';
import './menu/head-menu-props.mjs';
import './menu/submenu.mjs';
import './menu/submenu-props.mjs';
import './common-components/fake-arrow.mjs';
import './tabs/tabs.mjs';
import './tabs/tab-panel.mjs';
import './tabs/tab-panel-props.mjs';
import './tabs/tab-nav.mjs';
import './tabs/props.mjs';
import './tabs/tab-nav-item.mjs';
import './tabs/tab-nav-bar.mjs';
import './menu/menu-item.mjs';
import './menu/menu-item-props.mjs';
import './menu/menu-group.mjs';
import './menu/menu-group-props.mjs';
import './pagination/pagination.mjs';
import './pagination/props.mjs';
import './pagination/hooks/index.mjs';
import './pagination/hooks/useMoreAction.mjs';
import './pagination/hooks/usePaginationClasses.mjs';
import './input-number/input-number.mjs';
import './input-number/props.mjs';
import './input-number/hooks/useInputNumber.mjs';
import './_chunks/dep-dd322626.mjs';
import './_chunks/dep-550c1c04.mjs';
import './input/input.mjs';
import './input/props.mjs';
import './input/hooks/index.mjs';
import './input/hooks/useInput.mjs';
import './form/consts/index.mjs';
import './input/hooks/useLengthLimit.mjs';
import './input/hooks/useInputEventHandler.mjs';
import './input/hooks/useInputWidth.mjs';
import './input/input-group.mjs';
import './input/input-group-props.mjs';
import './select/select.mjs';
import './select/components/select-panel.mjs';
import './select/option.mjs';
import './select/option-props.mjs';
import './select/utils/index.mjs';
import './select/consts/index.mjs';
import './checkbox/checkbox.mjs';
import './checkbox/props.mjs';
import './checkbox/consts/index.mjs';
import './checkbox/hooks/useCheckboxLazyLoad.mjs';
import './checkbox/hooks/useKeyboardEvent.mjs';
import './checkbox/group.mjs';
import './checkbox/checkbox-group-props.mjs';
import './select/option-group.mjs';
import './select/option-group-props.mjs';
import './select/props.mjs';
import './select/hooks/index.mjs';
import './select/hooks/useKeyboardControl.mjs';
import './select/hooks/usePanelVirtualScroll.mjs';
import './select/hooks/useSelectOptions.mjs';
import './select-input/select-input.mjs';
import './select-input/props.mjs';
import './select-input/hooks/index.mjs';
import './select-input/hooks/useMultiple.mjs';
import './tag-input/tag-input.mjs';
import './tag-input/props.mjs';
import './tag-input/hooks/index.mjs';
import './tag-input/hooks/useDragSorter.mjs';
import './tag-input/hooks/useHover.mjs';
import './tag-input/hooks/useTagScroll.mjs';
import './tag-input/hooks/useTagList.mjs';
import './tag/tag.mjs';
import './_chunks/dep-e5707621.mjs';
import './tag/props.mjs';
import './tag/check-tag.mjs';
import './tag/check-tag-props.mjs';
import './tag/check-tag-group.mjs';
import './tag/check-tag-group-props.mjs';
import './select-input/hooks/useOverlayInnerStyle.mjs';
import './select-input/hooks/useSingle.mjs';
import './input-adornment/input-adornment.mjs';
import './input-adornment/props.mjs';
import './pagination/pagination-mini.mjs';
import './pagination/pagination-mini-props.mjs';
import './steps/steps.mjs';
import './steps/props.mjs';
import './steps/step-item-props.mjs';
import './steps/step-item.mjs';
import './sticky-tool/sticky-tool.mjs';
import './sticky-tool/props.mjs';
import './sticky-tool/sticky-item.mjs';
import './sticky-tool/sticky-item-props.mjs';
import './back-top/back-top.mjs';
import './back-top/props.mjs';
import './auto-complete/auto-complete.mjs';
import './auto-complete/props.mjs';
import './auto-complete/components/option-list.mjs';
import './auto-complete/components/highlight-option.mjs';
import './cascader/cascader.mjs';
import './cascader/components/Panel.mjs';
import './cascader/components/Item.mjs';
import './cascader/utils/index.mjs';
import './cascader/utils/className.mjs';
import './cascader/utils/effect.mjs';
import './cascader/utils/helper.mjs';
import './cascader/props.mjs';
import './cascader/hooks/index.mjs';
import './_chunks/dep-2318c8f8.mjs';
import 'mitt';
import './_chunks/dep-517a8ba2.mjs';
import './cascader/cascader-panel.mjs';
import './date-picker/DatePicker.mjs';
import './date-picker/hooks/useSingle.mjs';
import './_chunks/dep-94c99e54.mjs';
import './date-picker/hooks/useSingleValue.mjs';
import './_chunks/dep-5a18ab75.mjs';
import './_chunks/dep-f2a81bf0.mjs';
import './date-picker/props.mjs';
import './date-picker/components/panel/SinglePanel.mjs';
import './date-picker/components/panel/PanelContent.mjs';
import './date-picker/components/base/Header.mjs';
import './date-picker/components/base/Table.mjs';
import './date-picker/components/base/Cell.mjs';
import './time-picker/panel/time-picker-panel.mjs';
import './_chunks/dep-f80b6173.mjs';
import './time-picker/props.mjs';
import './_chunks/dep-75c7451f.mjs';
import './date-picker/components/panel/ExtraContent.mjs';
import './date-picker/components/base/Footer.mjs';
import './date-picker/hooks/index.mjs';
import './date-picker/hooks/useDisableDate.mjs';
import './date-picker/hooks/useRange.mjs';
import './date-picker/hooks/useRangeValue.mjs';
import './date-picker/hooks/useTableData.mjs';
import './date-picker/DatePickerPanel.mjs';
import './date-picker/date-picker-panel-props.mjs';
import './date-picker/DateRangePicker.mjs';
import './date-picker/date-range-picker-props.mjs';
import './date-picker/components/panel/RangePanel.mjs';
import './date-picker/utils/index.mjs';
import './range-input/range-input.mjs';
import './range-input/props.mjs';
import './range-input/range-input-popup.mjs';
import './range-input/range-input-popup-props.mjs';
import './date-picker/DateRangePickerPanel.mjs';
import './date-picker/date-range-picker-panel-props.mjs';
import './form/form.mjs';
import './form/props.mjs';
import './form/utils/index.mjs';
import './form/utils/form-model.mjs';
import './form/utils/form-item.mjs';
import './form/form-item.mjs';
import './form/form-item-props.mjs';
import './radio/radio.mjs';
import './radio/props.mjs';
import './radio/consts/index.mjs';
import './radio/group.mjs';
import './radio/radio-group-props.mjs';
import './radio/radio-button.mjs';
import './radio/hooks/index.mjs';
import './radio/hooks/useKeyboard.mjs';
import './watermark/hooks/index.mjs';
import './slider/slider.mjs';
import './slider/props.mjs';
import './slider/slider-button.mjs';
import './slider/hooks/useSliderTooltip.mjs';
import './slider/utils/index.mjs';
import './slider/consts/index.mjs';
import './slider/hooks/useSliderMark.mjs';
import './slider/slider-mark.mjs';
import './slider/hooks/useSliderInput.mjs';
import './switch/switch.mjs';
import './switch/props.mjs';
import './textarea/textarea.mjs';
import './textarea/props.mjs';
import './transfer/transfer.mjs';
import './transfer/components/transfer-list.mjs';
import './transfer/utils/index.mjs';
import './transfer/components/transfer-search.mjs';
import './transfer/hooks/useDragSort.mjs';
import './transfer/components/transfer-operations.mjs';
import './transfer/props.mjs';
import './time-picker/time-picker.mjs';
import './time-picker/time-range-picker.mjs';
import './time-picker/time-range-picker-props.mjs';
import './tree-select/tree-select.mjs';
import './tree-select/props.mjs';
import './tree/tree.mjs';
import './tree/utils/adapt.mjs';
import './tree/hooks/useDragHandle.mjs';
import './tree/hooks/useDraggable.mjs';
import './tree/utils/index.mjs';
import './tree/hooks/useTreeAction.mjs';
import './tree/hooks/useTreeNodes.mjs';
import './tree/tree-item.mjs';
import './tree/hooks/useItemState.mjs';
import './tree/hooks/useTreeItem.mjs';
import './tree/hooks/useItemEvents.mjs';
import './tree/hooks/useRenderIcon.mjs';
import './tree/hooks/useRenderLabel.mjs';
import './tree/hooks/useRenderLine.mjs';
import './tree/hooks/useRenderOperations.mjs';
import './tree/hooks/useTreeEvents.mjs';
import './tree/hooks/useTreeScroll.mjs';
import './tree/hooks/useTreeState.mjs';
import './tree/hooks/useTreeStore.mjs';
import './tree/hooks/useTreeStyles.mjs';
import './tree/props.mjs';
import './color-picker/color-picker-panel.mjs';
import './color-picker/color-picker-panel-props.mjs';
import './color-picker/components/panel/index.mjs';
import './_chunks/dep-baeee0c4.mjs';
import './_chunks/dep-102ee016.mjs';
import './color-picker/components/panel/header.mjs';
import './color-picker/hooks/index.mjs';
import './color-picker/components/panel/linear-gradient.mjs';
import './color-picker/utils/index.mjs';
import './color-picker/components/panel/base-props.mjs';
import './color-picker/components/panel/saturation.mjs';
import './color-picker/components/panel/hue.mjs';
import './color-picker/components/panel/slider.mjs';
import './color-picker/components/panel/alpha.mjs';
import './color-picker/components/format/index.mjs';
import './color-picker/props.mjs';
import './color-picker/components/format/inputs.mjs';
import './color-picker/components/panel/swatches.mjs';
import './color-picker/color-picker.mjs';
import './color-picker/components/trigger/index.mjs';
import './avatar/avatar.mjs';
import './avatar/props.mjs';
import './image/image.mjs';
import './image/props.mjs';
import './avatar/group.mjs';
import './avatar/avatar-group-props.mjs';
import './badge/badge.mjs';
import './badge/props.mjs';
import './calendar/calendar.mjs';
import './calendar/props.mjs';
import './calendar/utils/index.mjs';
import './calendar/consts/index.mjs';
import './calendar/hooks/index.mjs';
import './calendar/hooks/useState.mjs';
import './calendar/hooks/useCalendarClass.mjs';
import './calendar/hooks/useController.mjs';
import './calendar/hooks/useColHeaders.mjs';
import './calendar/calendar-cell.mjs';
import './card/card.mjs';
import './card/props.mjs';
import './comment/comment.mjs';
import './comment/props.mjs';
import './empty/empty.mjs';
import './empty/props.mjs';
import './empty/components/MaintenanceSvg.mjs';
import './empty/components/NetworkErrorSvg.mjs';
import './empty/components/EmptySvg.mjs';
import './empty/components/FailSvg.mjs';
import './empty/components/SuccessSvg.mjs';
import './image-viewer/image-viewer.mjs';
import './image-viewer/base/ImageItem.mjs';
import './image-viewer/hooks/index.mjs';
import './image-viewer/base/ImageModalIcon.mjs';
import './image-viewer/base/ImageViewerModal.mjs';
import './image-viewer/props.mjs';
import './image-viewer/base/ImageViewerUtils.mjs';
import './dialog/dialog.mjs';
import './dialog/props.mjs';
import './dialog/hooks/index.mjs';
import './dialog/hooks/useAction.mjs';
import './dialog/hooks/useSameTarget.mjs';
import './_chunks/dep-635a401b.mjs';
import './dialog/utils/index.mjs';
import './dialog/dialog-card.mjs';
import './dialog/dialog-card-props.mjs';
import './image-viewer/consts/index.mjs';
import './image-viewer/utils/index.mjs';
import './list/list.mjs';
import './list/list-item.mjs';
import './list/list-item-props.mjs';
import './list/props.mjs';
import './list/consts/index.mjs';
import './list/hooks/index.mjs';
import './list/hooks/useListItems.mjs';
import './list/hooks/useListVirtualScroll.mjs';
import './list/list-item-meta.mjs';
import './list/list-item-meta-props.mjs';
import './progress/progress.mjs';
import './progress/consts/index.mjs';
import './progress/props.mjs';
import './skeleton/skeleton.mjs';
import './skeleton/props.mjs';
import './statistic/statistic.mjs';
import './statistic/props.mjs';
import './swiper/swiper.mjs';
import './swiper/props.mjs';
import './swiper/swiper-item.mjs';
import './table/base-table.mjs';
import './table/base-table-props.mjs';
import './table/hooks/useTableHeader.mjs';
import './table/hooks/useMultiHeader.mjs';
import './table/hooks/useClassName.mjs';
import './table/components/ellipsis.mjs';
import './table/hooks/useColumnResize.mjs';
import './table/hooks/useFixed.mjs';
import './table/hooks/usePagination.mjs';
import './table/hooks/useAffix.mjs';
import './table/components/tbody.mjs';
import './table/components/tr.mjs';
import './table/utils/index.mjs';
import './table/hooks/useLazyLoad.mjs';
import './table/hooks/useRowspanAndColspan.mjs';
import './table/hooks/useStyle.mjs';
import './table/components/thead.mjs';
import './table/components/tfoot.mjs';
import './table/hooks/useRowHighlight.mjs';
import './table/hooks/useHoverKeyboardEvent.mjs';
import './table/primary-table.mjs';
import './table/primary-table-props.mjs';
import './table/hooks/useColumnController.mjs';
import './table/components/column-checkbox-group.mjs';
import './table/hooks/useRowExpand.mjs';
import './table/hooks/useRowSelect.mjs';
import './_chunks/dep-916153fa.mjs';
import './table/hooks/useSorter.mjs';
import './table/components/sorter-button.mjs';
import './table/hooks/useFilter.mjs';
import './table/components/filter-controller.mjs';
import './table/hooks/useDragSort.mjs';
import 'sortablejs';
import './table/hooks/useAsyncLoading.mjs';
import './table/components/editable-cell.mjs';
import './_chunks/dep-cc23cde4.mjs';
import './table/hooks/useEditableRow.mjs';
import './table/enhanced-table.mjs';
import './table/enhanced-table-props.mjs';
import './table/hooks/useTreeData.mjs';
import './_chunks/dep-5803e9fd.mjs';
import './table/hooks/useTreeSelect.mjs';
import './timeline/timeline.mjs';
import './timeline/props.mjs';
import './timeline/hooks/index.mjs';
import './timeline/timeline-item.mjs';
import './timeline/timeline-item-props.mjs';
import './collapse/collapse.mjs';
import './collapse/props.mjs';
import './collapse/collapse-panel.mjs';
import './collapse/collapse-panel-props.mjs';
import './watermark/watermark.mjs';
import './watermark/props.mjs';
import './rate/rate.mjs';
import './rate/props.mjs';
import './typography/typography.mjs';
import './typography/text.mjs';
import './typography/text-props.mjs';
import './typography/utils/copy-to-clipboard.mjs';
import './typography/components/ellipsis.mjs';
import './typography/paragraph-props.mjs';
import './typography/title.mjs';
import './typography/title-props.mjs';
import './typography/paragraph.mjs';
import './qrcode/qrcode.mjs';
import './qrcode/props.mjs';
import './_chunks/dep-4bf85467.mjs';
import './qrcode/components/qrcode-canvas.mjs';
import './qrcode/components/props.mjs';
import './qrcode/components/qrcode-svg.mjs';
import './qrcode/components/qrcode-status.mjs';
import './alert/alert.mjs';
import './alert/props.mjs';
import './drawer/drawer.mjs';
import './drawer/props.mjs';
import './drawer/hooks/index.mjs';
import './guide/guide.mjs';
import './guide/props.mjs';
import './guide/utils/index.mjs';
import './guide/utils/getScrollParent.mjs';
import './guide/utils/getRelativePosition.mjs';
import './guide/utils/getTargetElm.mjs';
import './guide/utils/scrollToElm.mjs';
import './notification/notification-list.mjs';
import './notification/notification.mjs';
import './notification/props.mjs';
import './notification/utils/index.mjs';
import './notification/utils/animate.mjs';
import './notification/consts/index.mjs';
import './popconfirm/popconfirm.mjs';
import './popconfirm/props.mjs';
import './upload/upload.mjs';
import './upload/props.mjs';
import './upload/components/normal-file.mjs';
import './upload/consts/index.mjs';
import './upload/components/dragger-file.mjs';
import './upload/hooks/useDrag.mjs';
import './upload/components/image-card.mjs';
import './upload/components/multiple-flow-list.mjs';
import './upload/components/custom-file.mjs';
import './upload/hooks/useUpload.mjs';
import './config-provider/config-provider.mjs';
import './config-provider/props.mjs';

function install(app, config) {
  Object.keys(components).forEach(function (key) {
    if (/directive/i.test(key)) return;
    /plugin/i.test(key) ? app.use(components[key]) : app.use(components[key], config);
  });
}
var index = {
  install: install,
  version: typeof "1.15.2" === "undefined" ? "" : "1.15.2"
};

export { index as default, install };
